### DISCUSSION TITLE
[@types/node-schedule] Job.schedule missing valid parameter types

### DISCUSSION THREAD

**Discussion by dever23b** [DefinitelyTyped/DefinitelyTyped#62062] [Issues with a @types package]
Hello:

***TL;DR*** I believe the parameter `date` on `Job.schedule` should accept the same parameter types as `rule` on `scheduleJob`, and perhaps be renamed to `rule` as well for the sake of clarity:
```typescript
rule: RecurrenceRule | RecurrenceSpecDateRange | RecurrenceSpecObjLit | Date | string | number
```
Rationale: The two parameters represent the same data, as it is passed directly from `scheduleJob` to `Job.schedule` without modification.

---

I am trying out `node-schedule` and there appears to be a discrepancy between the types defined in `@types/node-schedule` and `node-schedule` itself.  `@types/node-schedule`, as defined, restricts users to using the `scheduleJob` method from `node-schedule`; however, the documentation of `node-schedule` suggests that `scheduleJob` is simply a convenience method.  Attempting to create a `Job` instance directly and schedule its invocation results in a TypeScript compilation error if using a subset of the allowable parameter types.

It seems like `scheduleJob`'s `rule` (as defined in `@types/node-schedule`)/`spec` (as defined in `node-schedule`) parameter and `Job.schedule`'s `date` (`@types/node-schedule`)/`spec` (`node-schedule`) parameter should accept the same types; however, they are defined differently (and with different names, although this seems only to indicate that this is unintended rather than to have any functional difference) in `@types/node-schedule` and that creates a restriction.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/8e501a79dc4e7a8a03a4a2e9df8dd6443a3fe5dc/types/node-schedule/index.d.ts#L89-L90

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/8e501a79dc4e7a8a03a4a2e9df8dd6443a3fe5dc/types/node-schedule/index.d.ts#L181-L196

Justification:

#### ***[node-schedule's example](https://github.com/node-schedule/node-schedule#recurrence-rule-scheduling)***
```javascript
const schedule = require('node-schedule');

const rule = new schedule.RecurrenceRule();
rule.minute = 42;

const job = schedule.scheduleJob(rule, function(){
  console.log('The answer to life, the universe, and everything!');
});
```
[Behind the scenes](https://github.com/node-schedule/node-schedule/blob/017c195e3ababe9a7eb80c2565ef2d9162472e7b/lib/schedule.js#L23-L44), `schedule.scheduleJob` simply creates a `Job` instance and passes the `spec` parameter directly to `Job.schedule` without modification:
```javascript
function scheduleJob() {
  if (arguments.length < 2) {
    throw new RangeError('Invalid number of arguments');
  }


  const name = (arguments.length >= 3 && typeof arguments[0] === 'string') ? arguments[0] : null;
  const spec = name ? arguments[1] : arguments[0];
  const method = name ? arguments[2] : arguments[1];
  const callback = name ? arguments[3] : arguments[2];


  if (typeof method !== 'function') {
    throw new RangeError('The job method must be a function.');
  }


  const job = new Job(name, method, callback);


  if (job.schedule(spec)) {
    return job;
  }


  return null;
}
```
Based on this implementation, the `Job.schedule` method seems to be designed to receive any type that `scheduleJob` does.  [I verified it here](https://github.com/node-schedule/node-schedule/blob/017c195e3ababe9a7eb80c2565ef2d9162472e7b/lib/Job.js#L236-L262): indeed, the `schedule` method is designed to accept a `RecurrenceRule` or an object resembling that class.

`@types/node-schedule` allows for the example above (using `scheduleJob`) to compile; however, if I try to do exactly what `scheduleJob` does and create/schedule `Job` directly, it fails due to the declaration of `Job.schedule`:
```typescript
protected schedule(
    name: string,
    schedule: RecurrenceRule | RecurrenceSpecDateRange | RecurrenceSpecObjLit | Date | string | number,
    callback: JobCallback,
    runImmediately: boolean = false
  ): void {
    //...
    const job: Job = new Job(name, callback);
    if( !job.schedule(schedule) ) {
      throw new BaseException('Invalid schedule defined');
    }

    //... 
  }
```
This code fails at `job.schedule(schedule)` because the `@types/node-schedule` declaration for `Job.schedule` does not permit the `RecurrenceRule` associated types that `scheduleJob` accepts.

Therefore, I believe that the definition in `@types/node-schedule` contains a bug and humbly request it be adjusted to synchronize the two sets of allowable parameter types.

I hope this explains enough.  Thank you for your time!

[a comment by an automated account is not shown]

**Comment by dever23b** [comment]
Seems like the same issue applies to `Job.reschedule` as well
